core/src/utilities/helpers/routing-helpers.js

```diff
@@ -123,6 +100,42 @@ export const buildRoute = (node, path, params) =>
     ? path + (params ? '?' + params : '')
     : buildRoute(node.parent, `/${node.parent.pathSegment}${path}`, params);
 
+export const processDynamicNode = (node, pathParams) => {
```

Comment:
This should be eliminated as it mutates the given node.

Reply:
After this has been removed the way the navigation links are built must be reworked (routing.js handleRouteClick(node) must take the current navigation path from the URL into consideration to properly deal with dynamic path segments in the current navigation path; currently this function builds the navigation path from scratch based on node structure alone which won't work with unmutated navigation nodes any more).